scipy/optimize/zeros.py

```diff
@@ -206,6 +222,86 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     raise RuntimeError(msg)
 
 
+def _array_newton(func, x0, fprime, args, tol, maxiter, fprime2,
+                  failure_idx_flag=False):
+    """
+    A vectorized version of Newton, Halley, and secant methods for arrays. Do
+    not use this method directly. This method is called from :func:`newton`
+    when ``np.isscalar(x0)`` is true. For docstring, see :func:`newton`.
+    """
+    p = np.asarray(x0)
+    failures = np.ones_like(p, dtype=bool)  # at start, nothing converged
+    if fprime is not None:
+        # Newton-Raphson method
+        for iteration in range(maxiter):
+            fder = np.asarray(fprime(p, *args))
+            zero_der = (fder == 0)
+            # stop iterating if all derivatives are zero
+            if zero_der.all():
+                break
+            fval = np.asarray(func(p, *args))
+            # Newton step
+            dp = fval / fder
```

Comment:
We cannot have spurious divison by zero errors produced by our code, it is simply not an option. So you are going to have to limit the items for which you do this computation, e.g.:

```
dp = fval[~zero_der] / fder[~zero_der]
```

and similarly for the second derivative calculations.

I think there is a case here for making `nonzero_der` your mask.

If all this boolean indexing slows things down, consider checking if there actually is anything to mask out before doing the masking, i.e. by checking if `zero_der.any()` or `not nonzero_der.all()`. Although this is premature optimization, and you probably know what that is... ;-)

Reply:
✔️